### PR TITLE
Add Sentiment Analysis Middleware Component

### PR DESCRIPTION
**Pull request by arafattehsin**
# Sentiment Analysis Middleware Component for Bot Framework Composer

## Build status
| Branch | Status | Recommended NuGet package version |
| ------ | ------ | ------ |
| master | [Build status](https://ci.appveyor.com/project/garypretty/botbuilder-community) | [Available via NuGet](https://www.nuget.org/packages/Bot.Builder.Community.Components.Handoff.LivePerson/) |

## Description

This is part of the [Bot Builder Community](https://github.com/botbuildercommunity) project which contains open source Bot Framework Composer components, along with other extensions, including middleware, recognizers and other components for use with the Bot Builder .NET SDK v4.

This component enables the addition of the middleware component for Sentiment Analysis of each incoming message. This component can extract the sentiment powered by [SentimentAnalyzer](https://www.nuget.org/packages/SentimentAnalyzer/) which encapsulates both offline (powered by ML.NET) and online (powered by Azure Cognitive Services) sentiment analysis.

## Usage

* [Composer component installation](#composer-component-installation)
* [Composer component configuration](#composer-component-configuration)
* [Getting the Sentiment Analysis result from Middleware](#getting-the-sentiment-analysis-result-from-middleware)

### Composer component installation

1. Navigate to the Bot Framework Composer **Package Manager**.
2. Change the filter to **Community packages**.
3. Search for 'sentiment analysis' and install **Bot.Builder.Community.Components.Middleware.SentimentAnalysis**

image

### Composer component configuration

1. Within Composer, navigate to **Project Settings** and toggle the **Advanced Settings View (json)**.
2. Add the following settings at the root of your settings JSON, replacing the placeholders as described below.

```json
"Bot.Builder.Community.Components.Middleware.SentimentAnalysis": {
"IsEnabled": true,
"APIKey": "",
"EndpointUrl": ""
}
```
> [!NOTE]
> If you do not want to use the middleware, you can simply set the value of IsEnabled to false. The above setting will work with offline sentiment analysis (powered by Microsoft ML.NET). However, if you want to use the Azure Text Analytics service to find out the Sentiments then you may have to provide API Key and Endpoint URL as shown below;

image

### Getting the Sentiment Analysis result from Middleware

Once you've configured the middleware component and enabled it from the Connections (`"IsEnabled": true`) then this component will run on every message received (`ActivityType is Message`). 

You can now get the results using two different settings;

#### Offline Mode (without any external API) 

To use this, you just have to keep the settings as it is defined in the above example. In order to get it in Bot Framework Composer, you can use this syntax;

`${turn.Sentiment}`

image

This will give result as **True** for Positive and **False ** for Negative.

image

#### Online Mode (using Azure Text Analytics) 

To use this, you need to get the API key and an endpoint URL, once they both are populated, the component will automatically switch its mode to online and now the results will be fetched from Azure Text Analytics service. Same syntax as above;

`${turn.Sentiment}`

This will give result from 0.0 to 1 where 1 being Positive and 0 being Negative. 

image

`The result is 0.1 (Negative)`

### CODE REVIEW

**Comment by arafattehsin** [comment]
Hi @garypretty - this component is using 2 external NuGet packages. Just curious, is it something I'll have to see at my end or a new step of restoring the packages needs to be added in the pipeline?

File: libraries/Bot.Builder.Community.Components.Middleware.SentimentAnalysis/Models/SentimentAnalysisCredentialsProvider.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Bot.Builder.Community.Components.Middleware.SentimentAnalysis.Models
+{
+    public class SentimentAnalysisCredentialsProvider : ISentimentAnalysisCredentialsProvider
+    {
+        public SentimentAnalysisCredentialsProvider(IConfiguration configuration)
+        {
+            IsEnabled = Convert.ToBoolean(configuration["IsEnabled"]);
```
Comment (rvinothrajendran) [inline libraries/Bot.Builder.Community.Components.Middleware.SentimentAnalysis/Models/SentimentAnalysisCredentialsProvider.cs:12]: @arafattehsin Before reading the value , first we have to check if the key is present or not ?

Comment (arafattehsin) [reply libraries/Bot.Builder.Community.Components.Middleware.SentimentAnalysis/Models/SentimentAnalysisCredentialsProvider.cs:12]: Good point. I'll add this validation.

**Comment by arafattehsin** [comment]
Addresses #390

**Review by garypretty — APPROVED** [review]

**Comment by garypretty** [comment]
@arafattehsin could you resolve the conflicts here please? Thanks :)

**Pull request merged**